### PR TITLE
[Reland] Add migrator to upgrade gradle version when conflict with Android Studio bundled Java version is detected

### PR DESCRIPTION
**Pull request by gmackall**
This is an attempt to reland https://github.com/flutter/flutter/pull/124085.

Differences from this attempt and the last: 
1. Adds a check for null android studio versions and a test for this case.
2. Wraps the migrate code in a try-catch [per the suggestion here](https://github.com/flutter/flutter/pull/125728/files#r1181747899).

Old PR description:
This PR adds an android project migrator that checks the version of android studio and the version of gradle for conflicts, and upgrades to 7.4 if a conflict is detected. For more detail about the particular conflict, see https://github.com/flutter/flutter/issues/122376.

The PR also upgrades older gradle versions being used in integration testing to 7.4.

Fixes/related to: https://github.com/flutter/flutter/issues/122376 and https://github.com/flutter/flutter/issues/123636

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart [hunk 1]
```diff
@@ -77,6 +111,169 @@ tasks.register("clean", Delete) {
...
+          androidSdk: FakeAndroidSdk(javaVersion: '17'),
+        );
+        gradleWrapperPropertiesFile.writeAsStringSync(gradleWrapperToMigrate);
+        migration.migrate();
+        expect(bufferLogger.traceText, contains(androidStudioNotFound));
+        expect(gradleWrapperPropertiesFile.readAsStringSync(),
+            gradleWrapperToMigrate);
+      });
+
+      testWithoutContext('skipped if android studio version is less than flamingo', () {
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:189]: Can you make a test that exercises your try statement? 

For example having androidSdk.getJavaVersion throw an exception.

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart [hunk 2]
```diff
@@ -77,6 +111,169 @@ tasks.register("clean", Delete) {
...
+          processUtils: FakeProcessUtils(),
+          platform: FakePlatform(),
+          os: FakeOperatingSystemUtils(),
+          androidSdk: FakeAndroidSdk(javaVersion: '17'),
+        );
+        gradleWrapperPropertiesFile.writeAsStringSync(gradleWrapperToMigrate);
+        migration.migrate();
+        expect(gradleWrapperPropertiesFile.readAsStringSync(), gradleWrapperToMigrateTo);
+        expect(bufferLogger.statusText, contains('Conflict detected between Android Studio Java version and Gradle version, '
+            'upgrading Gradle version from 6.7 to 7.6.1.'));
```
Comment (reidbaker) [inline packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:257]: Consider using the constants defined here so if they upgrader uses a different version this test passes.

**Review by reidbaker — APPROVED** [review]

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart (hunk 2 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart]: I've changed to use the constant defined for version we upgrade to, 7.6.1 (the other one is parsed in the migration method).

File: packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart (hunk 1 above)
Comment (gmackall) [reply packages/flutter_tools/test/general.shard/android/android_project_migration_test.dart:207]: Added a test following the above suggestion, let me know if it looks good!

File: packages/flutter_tools/lib/src/android/migrations/android_studio_java_gradle_conflict_migration.dart [hunk 3]
```diff
@@ -20,7 +20,7 @@ import '../gradle_utils.dart';
 @visibleForTesting
 final Version androidStudioFlamingo = Version(2022, 2, 0);
 
-const String _gradleVersion7_6_1 = r'7.6.1';
+const String gradleVersion7_6_1 = r'7.6.1';
```
Comment (reidbaker) [inline packages/flutter_tools/lib/src/android/migrations/android_studio_java_gradle_conflict_migration.dart:23]: I think this should be annotated @visibleForTesting

**Review by reidbaker — APPROVED** [review]

File: dev/integration_tests/flutter_gallery/perf.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"targets":[{"name":"gen_localizations","skipped":true,"succeeded":true,"elapsedMilliseconds":1},{"name":"gen_dart_plugin_registrant","skipped":true,"succeeded":true,"elapsedMilliseconds":4},{"name":"kernel_snapshot","skipped":true,"succeeded":true,"elapsedMilliseconds":459},{"name":"android_aot","skipped":true,"succeeded":true,"elapsedMilliseconds":45},{"name":"aot_android_asset_bundle","skipped":true,"succeeded":true,"elapsedMilliseconds":545},{"name":"android_aot","skipped":true,"succeeded":true,"elapsedMilliseconds":41},{"name":"android_aot","skipped":true,"succeeded":true,"elapsedMilliseconds":55},{"name":"android_aot_bundle_release_android-arm","skipped":true,"succeeded":true,"elapsedMilliseconds":45},{"name":"android_aot_bundle_release_android-arm64","skipped":true,"succeeded":true,"elapsedMilliseconds":41},{"name":"android_aot_bundle_release_android-x64","skipped":true,"succeeded":true,"elapsedMilliseconds":41},{"name":"_composite","skipped":true,"succeeded":true,"elapsedMilliseconds":0}]}
```
Comment (reidbaker) [inline dev/integration_tests/flutter_gallery/perf.json:1]: What is this file and why was it added?

**Review by reidbaker — APPROVED** [review]

File: dev/integration_tests/flutter_gallery/perf.json (hunk 4 above)
Comment (gmackall) [reply dev/integration_tests/flutter_gallery/perf.json]: It should not have been added, it was a mistaken add from a different terminal window. Reverted and re-did the commit, goal of commit was to just add the one missed gradle-wrapper file. 

Confirmed with grep that all gradle-wrapper files are now using 7.3+.

**Comment by gmackall** [comment]
Removing autosubmit briefly to test an assumption about RegExp group counts.

**Comment by gmackall** [comment]
Since last approval, I wrapped two different values returned by [RegExpMatch](https://api.flutter.dev/flutter/dart-core/RegExpMatch-class.html) group(). 

The main reason is that it is possible in a case like
```
final RegExp r = RegExp(r'ab(c)?');
final String s = 'ab';
final RegExpMatch m = r.firstMatch(s);
```
for `m.groupCount` to be 1, but `m[0]` to be null. It isn't possible with the RegExp as written, but given that the regex is in a different file it would be good for this check to be resilient to changes.

**Review by camsim99 — APPROVED** [review]
This looks good to me!

**Pull request merged**